File: kdb-storage-memtable/src/commonMain/kotlin/dev/kdb/storage/memtable/MemTable.kt
package dev.kdb.storage.memtable

import dev.kdb.codec.KdbHash
import dev.kdb.storage.PlatformIoShim
import dev.kdb.storage.sstable.DefaultSsTableWriter
import dev.kdb.storage.sstable.LsmBlobStore
import dev.kdb.storage.sstable.SsTableHandle
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * What one memtable generation knows about a key. A null [value] is a tombstone - deliberately
 * distinguishable from "no entry at all", which [MemTable.lookup] signals by returning null.
 */
public data class MemTableEntry(val value: ByteArray?) {
    public val deleted: Boolean get() = value == null

    override fun equals(other: Any?): Boolean =
        this === other || (other is MemTableEntry && value.contentEquals(other.value))

    override fun hashCode(): Int = value?.contentHashCode() ?: 0
}

public interface MemTable {
    public suspend fun put(key: KdbHash, value: ByteArray)

    /**
     * Returns the value for [key], or null if the key is absent *or* tombstoned. Callers that need
     * to tell those two apart - anything that would otherwise fall through to an older generation -
     * must use [lookup] instead.
     */
    public suspend fun get(key: KdbHash): ByteArray?

    /**
     * Reports what this generation knows about [key]. A null return means it has never seen the
     * key and the caller should keep searching older generations; a [MemTableEntry] with
     * `deleted = true` is a tombstone, and older generations must NOT be consulted - reading
     * through a tombstone is exactly how a delete failed to shadow an entry already flushed to an
     * SSTable.
     */
    public suspend fun lookup(key: KdbHash): MemTableEntry?

    public suspend fun delete(key: KdbHash)
    public val sizeBytes: Long
}

public class SortedMemTable : MemTable {
    private val map = linkedMapOf<KdbHash, ByteArray?>()
    private var bytes = 0L

    // put/delete now net out the *previous* value's size against the new one - previously bytes
    // only ever grew (every put added value.size, delete added nothing but never subtracted
    // anything either), so sizeBytes drifted arbitrarily far above the table's real footprint
    // under any workload with overwrites or deletes. Currently dormant (nothing reads sizeBytes
    // to trigger a flush yet - see MemTableManager's doc comment on the still-missing automatic
    // flush trigger), but wrong regardless, and would misfire the moment such a trigger exists.
    override suspend fun put(key: KdbHash, value: ByteArray) {
        val old = map.put(key, value)
        bytes += value.size - (old?.size ?: 0)
    }

    override suspend fun get(key: KdbHash): ByteArray? = map[key]

    override suspend fun lookup(key: KdbHash): MemTableEntry? =
        if (map.containsKey(key)) MemTableEntry(map[key]) else null

    // A key deleted before it was ever written still joins the insertion order (LinkedHashMap
    // keeps it), so a later put of it is not silently dropped from the flush snapshot.
    override suspend fun delete(key: KdbHash) {
        val old = map.put(key, null)
        bytes -= old?.size ?: 0
    }

    override val sizeBytes: Long get() = bytes

    internal fun snapshotEntries(): List<Pair<KdbHash, ByteArray?>> = map.entries.map { it.key to it.value }
}

public class MemTableManager(
    private val namespaceId: String,
    private val ioShim: PlatformIoShim,
    private val blobStore: LsmBlobStore,
) {
    private val mutex = Mutex()
    private var active = SortedMemTable()
    private var pendingFlush: SortedMemTable? = null

    public suspend fun put(key: KdbHash, value: ByteArray) = mutex.withLock { active.put(key, value) }

    /**
     * Tombstones [key] in the active memtable. The tombstone hides any value in the generation
     * being flushed and in the blob store, and survives the flush: [flush] writes it into the
     * SSTable as a real delete marker ([dev.kdb.storage.sstable.BlockHandle.deleted]), so a delete
     * of an already-flushed key stays deleted. Mirrors Go's memtable.Manager.Delete, which this
     * type previously had no counterpart for at all.
     */
    public suspend fun delete(key: KdbHash): Unit = mutex.withLock { active.delete(key) }

    /**
     * Reads through the generations newest-first, stopping at the first one with an opinion about
     * [key] - a tombstone included. The old chain of `?:` fallbacks could not stop at a tombstone
     * (they are indistinguishable from "absent" once flattened to a null value), so a delete never
     * shadowed the SSTable underneath it.
     */
    public suspend fun get(key: KdbHash): ByteArray? =
        mutex.withLock {
            active.lookup(key)?.let { return@withLock it.value }
            pendingFlush?.lookup(key)?.let { return@withLock it.value }
            blobStore.lookup(key)?.let { return@withLock if (it.deleted) null else it.value }
            null
        }

    // Clears pendingFlush only *after* the SSTable write actually succeeds - previously cleared
    // it immediately after staging the writer, before writer.finish() (the call that can
    // actually fail: I/O errors, a full disk, ...) ever ran. A throw there used to lose the
    // whole generation silently: active had already been swapped to a fresh empty table, so by
    // the time the exception propagated, the flushed-but-not-yet-durable writes were reachable
    // from neither active, pendingFlush (already nulled), nor blobStore (finish() never
    // completed) - get() would report every one of them as simply absent. They're still not
    // durably in an SSTable if finish() throws (this doesn't add a retry mechanism), but they
    // remain visible via pendingFlush until whatever called flush() decides how to recover.
    public suspend fun flush(level: Int = 0): SsTableHandle? =
        mutex.withLock {
            val snap = active
            pendingFlush = snap
            active = SortedMemTable()
            val writer = DefaultSsTableWriter(ioShim, namespaceId, level)
            var count = 0
            for ((k, v) in snap.snapshotEntries()) {
                // Tombstones are written, not skipped. Skipping them is what made a delete of an
                // already-flushed key temporary: the tombstone lived only in the memtable, so the
                // next flush erased the only record that the key had been deleted, and the value
                // reappeared from the SSTable it had been flushed into earlier. The format now
                // carries a delete marker (BlockHandle.deleted) on both sides of the port.
                if (v != null) writer.put(k, v) else writer.delete(k)
                count++
            }
            if (count == 0) {
                pendingFlush = null
                return@withLock null
            }
            val handle = writer.finish()
            blobStore.addTable(handle)
            pendingFlush = null
            handle
        }
}

File: go/kdb/storage/memtable/memtable.go
package memtable

import (
	"sync"

	"github.com/limidus/kdb/go/kdb/codec"
	"github.com/limidus/kdb/go/kdb/storage"
	"github.com/limidus/kdb/go/kdb/storage/sstable"
)

// Table is an in-memory sorted blob map.
type Table interface {
	Put(key codec.Hash, value []byte)
	Get(key codec.Hash) []byte
	Lookup(key codec.Hash) (value []byte, deleted, found bool)
	Delete(key codec.Hash)
	SizeBytes() int64
}

// SortedTable is a linked-order memtable.
type SortedTable struct {
	mu    sync.Mutex
	order []codec.Hash
	vals  map[codec.Hash]slot
	bytes int64
}

// slot is one key's state in the table. A tombstone is deleted=true with a nil value, which
// is deliberately distinguishable from "no entry at all" - the whole point of Lookup.
type slot struct {
	value   []byte
	deleted bool
}

// NewSortedTable returns an empty memtable.
func NewSortedTable() *SortedTable {
	return &SortedTable{vals: make(map[codec.Hash]slot)}
}

// Put writes value for key. SizeBytes nets out any value being replaced (including a
// tombstone's zero) rather than only ever growing.
func (m *SortedTable) Put(key codec.Hash, value []byte) {
	m.mu.Lock()
	defer m.mu.Unlock()
	old, ok := m.vals[key]
	if !ok {
		m.order = append(m.order, key)
	}
	m.vals[key] = slot{value: value}
	m.bytes += int64(len(value)) - int64(len(old.value))
}

// Get returns the value for key, or nil if the key is absent *or* tombstoned. Callers that
// need to tell those two apart - anything that would otherwise fall through to an older
// generation - must use Lookup instead.
func (m *SortedTable) Get(key codec.Hash) []byte {
	value, _, _ := m.Lookup(key)
	return value
}

// Lookup reports what this table knows about key. found=false means the table has never seen
// the key and the caller should keep searching older generations. found=true with
// deleted=true is a tombstone: the key was deleted here and older generations must NOT be
// consulted - reading through a tombstone is exactly how a delete used to fail to shadow an
// entry already flushed to an SSTable.
func (m *SortedTable) Lookup(key codec.Hash) (value []byte, deleted, found bool) {
	m.mu.Lock()
	defer m.mu.Unlock()
	s, ok := m.vals[key]
	if !ok {
		return nil, false, false
	}
	if s.deleted {
		return nil, true, true
	}
	return append([]byte(nil), s.value...), false, true
}

// Delete tombstones key, subtracting any value it replaces from SizeBytes. A key deleted
// before it was ever written still joins the insertion order, so a later Put of it is not
// silently dropped from the flush snapshot.
func (m *SortedTable) Delete(key codec.Hash) {
	m.mu.Lock()
	defer m.mu.Unlock()
	old, ok := m.vals[key]
	if !ok {
		m.order = append(m.order, key)
	}
	m.vals[key] = slot{deleted: true}
	m.bytes -= int64(len(old.value))
}

func (m *SortedTable) SizeBytes() int64 {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.bytes
}

func (m *SortedTable) snapshotEntries() []entry {
	m.mu.Lock()
	defer m.mu.Unlock()
	out := make([]entry, 0, len(m.order))
	for _, k := range m.order {
		s := m.vals[k]
		out = append(out, entry{key: k, value: s.value, deleted: s.deleted})
	}
	return out
}

type entry struct {
	key     codec.Hash
	value   []byte
	deleted bool
}

var _ Table = (*SortedTable)(nil)

// Manager coordinates active and flushing memtables with LSM blob storage.
type Manager struct {
	namespaceID string
	io          storage.PlatformIOShim
	blobStore   *sstable.LsmBlobStore

	mu           sync.Mutex
	active       *SortedTable
	pendingFlush *SortedTable
}

// NewManager returns a memtable manager for a namespace.
func NewManager(namespaceID string, io storage.PlatformIOShim, blobStore *sstable.LsmBlobStore) *Manager {
	return &Manager{
		namespaceID: namespaceID,
		io:          io,
		blobStore:   blobStore,
		active:      NewSortedTable(),
	}
}

func (mgr *Manager) Put(key codec.Hash, value []byte) {
	mgr.mu.Lock()
	defer mgr.mu.Unlock()
	mgr.active.Put(key, value)
}

// Delete tombstones key in the active memtable. The tombstone hides any value in the
// generation being flushed and in the blob store, but is dropped at the next flush: the
// SSTable format has no deleted marker, so a delete of a key that is already in an SSTable
// only holds for as long as the tombstone lives in memory. Persisting tombstones needs an
// SSTable format change, which per docs/go-porting.md has to originate on the Kotlin side.
func (mgr *Manager) Delete(key codec.Hash) {
	mgr.mu.Lock()
	defer mgr.mu.Unlock()
	mgr.active.Delete(key)
}

func (mgr *Manager) Get(key codec.Hash) []byte {
	mgr.mu.Lock()
	active := mgr.active
	pending := mgr.pendingFlush
	mgr.mu.Unlock()
	if v, deleted, found := active.Lookup(key); found {
		if deleted {
			return nil
		}
		return v
	}
	if pending != nil {
		if v, deleted, found := pending.Lookup(key); found {
			if deleted {
				return nil
			}
			return v
		}
	}
	return mgr.blobStore.Get(key)
}

// Flush writes the active memtable out as an SSTable and starts a fresh one. pendingFlush is
// cleared only once the write has actually succeeded: clearing it before Finish (the call
// that can fail on I/O) left the flushed-but-not-yet-durable generation reachable from
// neither active (already swapped), pendingFlush, nor the blob store (AddTable never
// reached), so Get reported every write in it as simply absent. On failure the generation
// stays visible through pendingFlush until the caller decides how to recover; it is still not
// durable, and a subsequent successful flush replaces pendingFlush with its own snapshot.
// Mirrors MemTableManager.flush in kdb-storage-memtable.
func (mgr *Manager) Flush(level int) (sstable.Handle, error) {
	mgr.mu.Lock()
	snap := mgr.active
	mgr.pendingFlush = snap
	mgr.active = NewSortedTable()
	mgr.mu.Unlock()

	writer := sstable.NewDefaultWriter(mgr.io, mgr.namespaceID, level)
	count := 0
	for _, e := range snap.snapshotEntries() {
		if e.deleted {
			continue
		}
		writer.Put(e.key, e.value)
		count++
	}
	if count == 0 {
		mgr.clearPending(snap)
		return sstable.Handle{}, nil
	}
	handle, err := writer.Finish()
	if err != nil {
		return sstable.Handle{}, err
	}
	mgr.blobStore.AddTable(handle)
	mgr.clearPending(snap)
	return handle, nil
}

// clearPending drops the pending-flush generation, but only if it is still the one this flush
// staged - a later flush may already have replaced it.
func (mgr *Manager) clearPending(snap *SortedTable) {
	mgr.mu.Lock()
	defer mgr.mu.Unlock()
	if mgr.pendingFlush == snap {
		mgr.pendingFlush = nil
	}
}
